utils/global-state-update-gen/src/generic.rs

```diff
@@ -0,0 +1,284 @@
+            recipients.remove(&public_key);
+        }
+    }
+
+    snapshot
+}
+
+/// Generates a set of writes necessary to "fix" the bids, ie.:
+/// - set the bids of the new validators to their desired stakes,
+/// - remove the bids of the old validators that are no longer validators,
+/// - remove all the bids that are larger than the smallest bid among the new validators
+/// (necessary, because such bidders would outbid the validators decided by the social
+/// consensus).
+pub fn add_and_remove_bids<T: StateReader>(
+    state: &mut StateTracker<T>,
+    validators_diff: &ValidatorsDiff,
+    new_snapshot: &SeigniorageRecipientsSnapshot,
+    only_listed_validators: bool,
+) {
+    let to_unbid = if only_listed_validators {
+        let large_bids = find_large_bids(state, new_snapshot);
+        validators_diff
+            .removed
+            .union(&large_bids)
+            .cloned()
+            .collect()
+    } else {
+        validators_diff.removed.clone()
+    };
+
+    for (pub_key, seigniorage_recipient) in new_snapshot.values().next().unwrap() {
+        let stake = *seigniorage_recipient.stake();
+        create_or_update_bid(state, pub_key, stake);
+    }
+
+    // Refresh the bids - we modified them above.
+    let bids = state.get_bids();
+
+    for pub_key in to_unbid {
+        if let Some(bid) = bids.get(&pub_key) {
+            let new_bid = Bid::empty(pub_key.clone(), *bid.bonding_purse());
+            state.set_bid(pub_key.clone(), new_bid);
+        }
+    }
+}
+
+/// Returns the set of public keys that have bids larger than the smallest bid among the new
+/// validators.
+fn find_large_bids<T: StateReader>(
+    state: &mut StateTracker<T>,
+    snapshot: &SeigniorageRecipientsSnapshot,
+) -> BTreeSet<PublicKey> {
+    let seigniorage_recipients = snapshot.values().next().unwrap();
+    let min_bid = seigniorage_recipients
+        .values()
+        .map(SeigniorageRecipient::stake)
+        .min()
+        .unwrap();
+    state
+        .get_bids()
+        .into_iter()
+        .filter(|(pub_key, bid)| {
+            bid.total_staked_amount()
+                .map_or(true, |amount| amount >= *min_bid)
+                && !seigniorage_recipients.contains_key(pub_key)
+        })
+        .map(|(pub_key, _bid)| pub_key)
+        .collect()
+}
+
+/// Updates the amount of an existing bid for the given public key, or creates a new one.
+fn create_or_update_bid<T: StateReader>(
+    state: &mut StateTracker<T>,
+    pub_key: &PublicKey,
+    stake: U512,
+) {
+    if state
+        .get_bids()
+        .get(pub_key)
+        .and_then(|bid| bid.total_staked_amount().ok())
+        == Some(stake)
+    {
+        // already staked the amount we need, nothing to do
+        return;
+    }
+    let new_bid = if let Some(bid) = state.get_bids().get(pub_key) {
+        Bid::unlocked(
+            pub_key.clone(),
+            *bid.bonding_purse(),
+            stake,
+            Default::default(),
```

Comment:
Does this set the delegation rate to 0?

But it doesn't remove the delegators, so in two eras, the validator's weight might increase again?

Reply:
Hmm, I'm not sure, but you might be right. Probably needs a closer look.